eliminate time.Sleep from server/comms/comms_test.go

We should plan to refactor the `comms` tests to eliminate sleeps by making use of the `wsConnStub.recv chan` that was added in https://github.com/decred/dcrdex/pull/287/commits/30fec9c817f7f9c575261a443f84a72233f177e2 for interested tests in to know when `(*wsConnStub).WriteMessage` did it's thing.

`time.Sleep` is quick but dirty unfortunately and CI apparently loves to preempt the test thread's CPU, breaking our assumptions about goroutine execution timing.

e.g.
https://github.com/decred/dcrdex/runs/694068914#step:5:92
have also seen `comms_test.go:427: error decoding last message (): unexpected end of JSON input`

This one doesn't seem related to sleeps, but I'll drop it here since you're working on the comms test:

```
panic: test timed out after 10m0s

goroutine 13 [running]:
testing.(*M).startAlarm.func1()
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:1459 +0x11c
created by time.goFunc
	/opt/hostedtoolcache/go/1.14.3/x64/src/time/sleep.go:168 +0x52

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000132c60, 0xb07c88, 0x12, 0xb234f8, 0x1)
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:1043 +0x699
testing.runTests.func1(0xc000132c60)
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:1284 +0xa7
testing.tRunner(0xc000132c60, 0xc000163cb0)
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:991 +0x1ec
testing.runTests(0xc00000e4a0, 0xf47380, 0x6, 0x6, 0x0)
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:1282 +0x528
testing.(*M).Run(0xc000164180, 0x0)
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:1199 +0x300
decred.org/dcrdex/server/comms.TestMain(0xc000164180)
	/home/runner/work/dcrdex/dcrdex/server/comms/comms_test.go:209 +0x100
main.main()
	_testmain.go:52 +0x224

goroutine 9 [chan send]:
decred.org/dcrdex/server/comms.sendReplace(0xc000132fc0, 0xc000072280, 0xa99da0, 0xc000020e00, 0xb0341b, 0x9, 0xafe190, 0x1)
	/home/runner/work/dcrdex/dcrdex/server/comms/comms_test.go:177 +0x1f7
decred.org/dcrdex/server/comms.TestClientRequests.func12(0xb0341b, 0x9, 0xafe190, 0x1)
	/home/runner/work/dcrdex/dcrdex/server/comms/comms_test.go:343 +0x1ad
decred.org/dcrdex/server/comms.TestClientRequests(0xc000132fc0)
	/home/runner/work/dcrdex/dcrdex/server/comms/comms_test.go:358 +0x904
testing.tRunner(0xc000132fc0, 0xb234f8)
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:991 +0x1ec
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.14.3/x64/src/testing/testing.go:1042 +0x661
FAIL	decred.org/dcrdex/server/comms	600.020s
```

--- a/client/asset/btc/btc.go
+++ b/client/asset/btc/btc.go
@@ -344,24 +344,24 @@ func (btc *ExchangeWallet) Info() *asset.WalletInfo {
 
 // Connect connects the wallet to the RPC server. Satisfies the dex.Connector
 // interface.
-func (btc *ExchangeWallet) Connect(ctx context.Context) (error, *sync.WaitGroup) {
+func (btc *ExchangeWallet) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	// Check the version. Do it here, so we can also diagnose a bad connection.
 	netVer, codeVer, err := btc.getVersion()
 	if err != nil {
-		return fmt.Errorf("error getting version: %v", err), nil
+		return nil, fmt.Errorf("error getting version: %v", err)
 	}
 	if netVer < btc.minNetworkVersion {
-		return fmt.Errorf("reported node version %d is less than minimum %d", netVer, minNetworkVersion), nil
+		return nil, fmt.Errorf("reported node version %d is less than minimum %d", netVer, minNetworkVersion)
 	}
 	if codeVer < minProtocolVersion {
-		return fmt.Errorf("node software out of date. version %d is less than minimum %d", codeVer, minProtocolVersion), nil
+		return nil, fmt.Errorf("node software out of date. version %d is less than minimum %d", codeVer, minProtocolVersion)
 	}
 	// If this is the first time connecting, clear the locked coins. This should
 	// have been done at shutdown, but shutdown may not have been clean.
 	if atomic.SwapUint32(&btc.hasConnected, 1) == 0 {
 		err := btc.wallet.LockUnspent(true, nil)
 		if err != nil {
-			return err, nil
+			return nil, err
 		}
 	}
 	var wg sync.WaitGroup
@@ -374,7 +374,7 @@ func (btc *ExchangeWallet) Connect(ctx context.Context) (error, *sync.WaitGroup)
 			btc.log.Errorf("failed to unlock %s outputs on shutdown: %v", btc.symbol, err)
 		}
 	}()
-	return nil, &wg
+	return &wg, nil
 }
 
 // Balance should return the total available funds in the wallet. Balance takes

--- a/client/asset/dcr/dcr.go
+++ b/client/asset/dcr/dcr.go
@@ -327,26 +327,26 @@ func (dcr *ExchangeWallet) Info() *asset.WalletInfo {
 
 // Connect connects the wallet to the RPC server. Satisfies the dex.Connector
 // interface.
-func (dcr *ExchangeWallet) Connect(ctx context.Context) (error, *sync.WaitGroup) {
+func (dcr *ExchangeWallet) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	err := dcr.client.Connect(ctx, false)
 	if err != nil {
-		return fmt.Errorf("Decred Wallet connect error: %v", err), nil
+		return nil, fmt.Errorf("Decred Wallet connect error: %v", err)
 	}
 	// Check the min api versions.
 	versions, err := dcr.client.Version()
 	if err != nil {
-		return fmt.Errorf("DCR ExchangeWallet version fetch error: %v", err), nil
+		return nil, fmt.Errorf("DCR ExchangeWallet version fetch error: %v", err)
 	}
 	err = checkVersionInfo(versions)
 	if err != nil {
-		return fmt.Errorf("DCR ExchangeWallet version check failed: %v", err), nil
+		return nil, fmt.Errorf("DCR ExchangeWallet version check failed: %v", err)
 	}
 	// If this is the first time connecting, clear the locked coins. This should
 	// have been done at shutdown, but shutdown may not have been clean.
 	if atomic.SwapUint32(&dcr.hasConnected, 1) == 0 {
 		err := dcr.node.LockUnspent(true, nil)
 		if err != nil {
-			return err, nil
+			return nil, err
 		}
 	}
 	var wg sync.WaitGroup
@@ -356,7 +356,7 @@ func (dcr *ExchangeWallet) Connect(ctx context.Context) (error, *sync.WaitGroup)
 		dcr.monitorBlocks(ctx)
 		dcr.shutdown()
 	}()
-	return nil, &wg
+	return &wg, nil
 }
 
 // Balance should return the total available funds in the wallet.

--- a/client/cmd/dexc/main.go
+++ b/client/cmd/dexc/main.go
@@ -17,6 +17,7 @@ import (
 	"decred.org/dcrdex/client/core"
 	"decred.org/dcrdex/client/rpcserver"
 	"decred.org/dcrdex/client/webserver"
+	"decred.org/dcrdex/dex"
 	"github.com/decred/slog"
 )
 
@@ -94,7 +95,13 @@ func main() {
 				log.Errorf("Error starting rpc server: %v", err)
 				os.Exit(1)
 			}
-			rpcSrv.Run(appCtx)
+			cm := dex.NewConnectionMaster(rpcSrv)
+			err = cm.Connect(appCtx)
+			if err != nil {
+				log.Errorf("Error starting rpc server: %v", err)
+				return
+			}
+			cm.Wait()
 		}()
 	}
 
@@ -104,10 +111,16 @@ func main() {
 			defer wg.Done()
 			webSrv, err := webserver.New(clientCore, cfg.WebAddr, logMaker.Logger("WEB"), cfg.ReloadHTML)
 			if err != nil {
-				log.Errorf("Error starting web server: %v", err)
+				log.Errorf("Error creating web server: %v", err)
 				os.Exit(1)
 			}
-			webSrv.Run(appCtx)
+			cm := dex.NewConnectionMaster(webSrv)
+			err = cm.Connect(appCtx)
+			if err != nil {
+				log.Errorf("Error starting web server: %v", err)
+				return
+			}
+			cm.Wait()
 		}()
 	}
 

--- a/client/cmd/dexc/ui/widgets.go
+++ b/client/cmd/dexc/ui/widgets.go
@@ -12,6 +12,7 @@ import (
 	"decred.org/dcrdex/client/core"
 	"decred.org/dcrdex/client/rpcserver"
 	"decred.org/dcrdex/client/webserver"
+	"decred.org/dcrdex/dex"
 	"github.com/decred/slog"
 	"github.com/gdamore/tcell"
 	"github.com/rivo/tview"
@@ -155,16 +156,23 @@ func createWidgets() {
 	marketView = newMarketView()
 	webView = newServerView("Web", cfg.WebAddr, func(ctx context.Context, addr string, logger slog.Logger) {
 		setWebLabelOn(true)
+		defer setWebLabelOn(false)
 		webSrv, err := webserver.New(clientCore, cfg.WebAddr, logger, cfg.ReloadHTML)
+		if err != nil {
+			log.Errorf("Error creating web server: %v", err)
+			return
+		}
+		cm := dex.NewConnectionMaster(webSrv)
+		err = cm.Connect(ctx)
 		if err != nil {
 			log.Errorf("Error starting web server: %v", err)
 			return
 		}
-		webSrv.Run(ctx)
-		setWebLabelOn(false)
+		cm.Wait()
 	})
 	rpcView = newServerView("RPC", cfg.RPCAddr, func(ctx context.Context, _ string, logger slog.Logger) {
 		setRPCLabelOn(true)
+		defer setRPCLabelOn(false)
 		rpcserver.SetLogger(logger)
 		rpcCfg := &rpcserver.Config{clientCore, cfg.RPCAddr, cfg.RPCUser, cfg.RPCPass, cfg.RPCCert, cfg.RPCKey}
 		rpcSrv, err := rpcserver.New(rpcCfg)
@@ -172,8 +180,14 @@ func createWidgets() {
 			log.Errorf("Error starting rpc server: %v", err)
 			return
 		}
-		rpcSrv.Run(ctx)
-		setRPCLabelOn(false)
+		cm := dex.NewConnectionMaster(rpcSrv)
+		err = cm.Connect(ctx)
+		if err != nil {
+			log.Errorf("Error starting rpc server: %v", err)
+			return
+		}
+		cm.Wait()
+
 	})
 	noteJournal = newJournal("Notifications", handleNotificationLog)
 	noteLog = mustLogger("NOTE", func(msg []byte) {

--- a/client/comms/wsconn.go
+++ b/client/comms/wsconn.go
@@ -47,7 +47,7 @@ type WsConn interface {
 	NextID() uint64
 	Send(msg *msgjson.Message) error
 	Request(msg *msgjson.Message, f func(*msgjson.Message)) error
-	Connect(ctx context.Context) (error, *sync.WaitGroup)
+	Connect(ctx context.Context) (*sync.WaitGroup, error)
 	MessageSource() <-chan *msgjson.Message
 }
 
@@ -375,7 +375,7 @@ func (conn *wsConn) NextID() uint64 {
 // connection will be returned. If the connection is successful, an
 // auto-reconnect goroutine will be started. To shutdown auto-reconnect, use
 // Stop() or cancel the context.
-func (conn *wsConn) Connect(ctx context.Context) (error, *sync.WaitGroup) {
+func (conn *wsConn) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	var ctxInternal context.Context
 	ctxInternal, conn.cancel = context.WithCancel(ctx)
 
@@ -400,7 +400,7 @@ func (conn *wsConn) Connect(ctx context.Context) (error, *sync.WaitGroup) {
 		close(conn.readCh) // signal to receivers that the wsConn is dead
 	}()
 
-	return conn.connect(ctxInternal), &conn.wg
+	return &conn.wg, conn.connect(ctxInternal)
 }
 
 // Stop can be used to close the connection and all of the goroutines started by

--- a/client/core/core_test.go
+++ b/client/core/core_test.go
@@ -218,8 +218,8 @@ func (conn *TWebsocket) Request(msg *msgjson.Message, f msgFunc) error {
 	return conn.reqErr
 }
 func (conn *TWebsocket) MessageSource() <-chan *msgjson.Message { return conn.msgs }
-func (conn *TWebsocket) Connect(context.Context) (error, *sync.WaitGroup) {
-	return conn.connectErr, &sync.WaitGroup{}
+func (conn *TWebsocket) Connect(context.Context) (*sync.WaitGroup, error) {
+	return &sync.WaitGroup{}, conn.connectErr
 }
 
 type TDB struct {
@@ -441,8 +441,8 @@ func (w *TXCWallet) Info() *asset.WalletInfo {
 	return &asset.WalletInfo{}
 }
 
-func (w *TXCWallet) Connect(ctx context.Context) (error, *sync.WaitGroup) {
-	return w.connectErr, &sync.WaitGroup{}
+func (w *TXCWallet) Connect(ctx context.Context) (*sync.WaitGroup, error) {
+	return &sync.WaitGroup{}, w.connectErr
 }
 
 func (w *TXCWallet) Run(ctx context.Context) { <-ctx.Done() }

--- a/client/rpcserver/rpcserver.go
+++ b/client/rpcserver/rpcserver.go
@@ -270,14 +270,14 @@ func New(cfg *Config) (*RPCServer, error) {
 }
 
 // Connect starts the RPC server. Satisfies the dex.Connector interface.
-func (s *RPCServer) Connect(ctx context.Context) (error, *sync.WaitGroup) {
+func (s *RPCServer) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	// ctx passed to newMarketSyncer when making new market syncers.
 	s.ctx = ctx
 
 	// Create listener.
 	listener, err := tls.Listen("tcp", s.addr, s.tlsConfig)
 	if err != nil {
-		return fmt.Errorf("can't listen on %s. rpc server quitting: %v", s.addr, err), nil
+		return nil, fmt.Errorf("can't listen on %s. rpc server quitting: %v", s.addr, err)
 	}
 
 	// Close the listener on context cancellation.
@@ -306,7 +306,7 @@ func (s *RPCServer) Connect(ctx context.Context) (error, *sync.WaitGroup) {
 		log.Infof("RPC server off")
 	}()
 	log.Infof("RPC server listening on %s", s.addr)
-	return nil, &s.wg
+	return &s.wg, nil
 }
 
 // handleRequest sends the request to the correct handler function if able.

--- a/client/rpcserver/rpcserver.go
+++ b/client/rpcserver/rpcserver.go
@@ -269,16 +269,15 @@ func New(cfg *Config) (*RPCServer, error) {
 	return s, nil
 }
 
-// Run starts the web server. Satisfies the dex.Runner interface.
-func (s *RPCServer) Run(ctx context.Context) {
+// Connect starts the RPC server. Satisfies the dex.Connector interface.
+func (s *RPCServer) Connect(ctx context.Context) (error, *sync.WaitGroup) {
 	// ctx passed to newMarketSyncer when making new market syncers.
 	s.ctx = ctx
 
 	// Create listener.
 	listener, err := tls.Listen("tcp", s.addr, s.tlsConfig)
 	if err != nil {
-		log.Errorf("can't listen on %s. rpc server quitting: %v", s.addr, err)
-		return
+		return fmt.Errorf("can't listen on %s. rpc server quitting: %v", s.addr, err), nil
 	}
 
 	// Close the listener on context cancellation.
@@ -292,19 +291,22 @@ func (s *RPCServer) Run(ctx context.Context) {
 			log.Errorf("HTTP server Shutdown: %v", err)
 		}
 	}()
-	log.Infof("RPC server listening on %s", s.addr)
-	if err := s.srv.Serve(listener); err != http.ErrServerClosed {
-		log.Warnf("unexpected (http.Server).Serve error: %v", err)
-	}
-	s.mtx.Lock()
-	for _, cl := range s.clients {
-		cl.Disconnect()
-	}
-	s.mtx.Unlock()
 
-	// Wait for market syncers to finish and Shutdown.
-	s.wg.Wait()
-	log.Infof("RPC server off")
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		if err := s.srv.Serve(listener); err != http.ErrServerClosed {
+			log.Warnf("unexpected (http.Server).Serve error: %v", err)
+		}
+		s.mtx.Lock()
+		for _, cl := range s.clients {
+			cl.Disconnect()
+		}
+		s.mtx.Unlock()
+		log.Infof("RPC server off")
+	}()
+	log.Infof("RPC server listening on %s", s.addr)
+	return nil, &s.wg
 }
 
 // handleRequest sends the request to the correct handler function if able.

--- a/client/rpcserver/rpcserver_test.go
+++ b/client/rpcserver/rpcserver_test.go
@@ -287,7 +287,7 @@ func TestLoadMarket(t *testing.T) {
 	link := newLink()
 	s, tCore, shutdown := newTServer(t, false, "", "")
 	defer shutdown()
-	err, _ := link.cl.Connect(tCtx)
+	_, err := link.cl.Connect(tCtx)
 	if err != nil {
 		t.Fatalf("WSLink Start: %v", err)
 	}

--- a/client/rpcserver/rpcserver_test.go
+++ b/client/rpcserver/rpcserver_test.go
@@ -235,11 +235,14 @@ func newTServer(t *testing.T, start bool, user, pass string) (*RPCServer,
 		t.Fatalf("error creating server: %v", err)
 	}
 	if start {
-		waiter := dex.NewStartStopWaiter(s)
-		waiter.Start(ctx)
+		cm := dex.NewConnectionMaster(s)
+		err := cm.Connect(ctx)
+		if err != nil {
+			t.Fatalf("Error starting WebServer: %v", err)
+		}
 		shutdown = func() {
 			killCtx()
-			waiter.WaitForShutdown()
+			cm.Disconnect()
 		}
 	} else {
 		shutdown = killCtx
@@ -284,7 +287,10 @@ func TestLoadMarket(t *testing.T) {
 	link := newLink()
 	s, tCore, shutdown := newTServer(t, false, "", "")
 	defer shutdown()
-	link.cl.Start()
+	err, _ := link.cl.Connect(tCtx)
+	if err != nil {
+		t.Fatalf("WSLink Start: %v", err)
+	}
 	defer link.cl.Disconnect()
 	params := &marketLoad{
 		Host:  "abc",

--- a/client/rpcserver/websocket.go
+++ b/client/rpcserver/websocket.go
@@ -92,8 +92,13 @@ func (s *RPCServer) websocketHandler(conn ws.Connection, ip string) {
 		delete(s.clients, cl.cid)
 		s.mtx.Unlock()
 	}()
-	cl.Start()
-	cl.WaitForShutdown()
+	cm := dex.NewConnectionMaster(cl)
+	err := cm.Connect(s.ctx)
+	if err != nil {
+		log.Errorf("websocketHandler client Connect: %v")
+		return
+	}
+	cm.Wait()
 	log.Tracef("Disconnected websocket client %s", ip)
 }
 

--- a/client/webserver/live_test.go
+++ b/client/webserver/live_test.go
@@ -707,7 +707,6 @@ func TestServer(t *testing.T) {
 	time.AfterFunc(time.Minute*59, func() { shutdown() })
 	logger := slog.NewBackend(os.Stdout).Logger("TEST")
 	logger.SetLevel(slog.LevelTrace)
-	time.AfterFunc(time.Minute*60, func() { shutdown() })
 	tCore := newTCore()
 
 	if register {
@@ -719,7 +718,11 @@ func TestServer(t *testing.T) {
 	if err != nil {
 		t.Fatalf("error creating server: %v", err)
 	}
-	go s.Run(tCtx)
+	cm := dex.NewConnectionMaster(s)
+	err = cm.Connect(tCtx)
+	if err != nil {
+		t.Fatalf("Connect error: %v", err)
+	}
 	go tCore.runEpochs()
-	<-tCtx.Done()
+	cm.Wait()
 }

--- a/client/webserver/webserver.go
+++ b/client/webserver/webserver.go
@@ -259,15 +259,14 @@ func New(core clientCore, addr string, logger slog.Logger, reloadHTML bool) (*We
 	return s, nil
 }
 
-// Run starts the web server. Satisfies the dex.Runner interface.
-func (s *WebServer) Run(ctx context.Context) {
+// Connect starts the web server. Satisfies the dex.Connector interface.
+func (s *WebServer) Connect(ctx context.Context) (error, *sync.WaitGroup) {
 	// We'll use the context for market syncers.
 	s.ctx = ctx
 	// Start serving.
 	listener, err := net.Listen("tcp", s.addr)
 	if err != nil {
-		log.Errorf("Can't listen on %s. web server quitting: %v", s.addr, err)
-		return
+		return fmt.Errorf("Can't listen on %s. web server quitting: %v", s.addr, err), nil
 	}
 
 	// Shutdown the server on context cancellation.
@@ -285,25 +284,28 @@ func (s *WebServer) Run(ctx context.Context) {
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
+		err = s.srv.Serve(listener)
+		if !errors.Is(err, http.ErrServerClosed) {
+			log.Warnf("unexpected (http.Server).Serve error: %v", err)
+		}
+		log.Infof("Web server off")
+		// Disconnect the websocket clients since Shutdown does not deal with
+		// hijacked websocket connections.
+		s.mtx.Lock()
+		for _, cl := range s.clients {
+			cl.Disconnect()
+		}
+		s.mtx.Unlock()
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
 		s.readNotifications(ctx)
 	}()
 
 	log.Infof("Web server listening on http://%s", s.addr)
-	err = s.srv.Serve(listener)
-	if !errors.Is(err, http.ErrServerClosed) {
-		log.Warnf("unexpected (http.Server).Serve error: %v", err)
-	}
-	log.Infof("Web server off")
-
-	// Disconnect the websocket clients since Shutdown does not deal with
-	// hijacked websocket connections.
-	s.mtx.Lock()
-	for _, cl := range s.clients {
-		cl.Disconnect()
-	}
-	s.mtx.Unlock()
-
-	wg.Wait()
+	return nil, &wg
 }
 
 // authorize creates, stores, and returns a new auth token to identify the

--- a/client/webserver/webserver.go
+++ b/client/webserver/webserver.go
@@ -260,13 +260,13 @@ func New(core clientCore, addr string, logger slog.Logger, reloadHTML bool) (*We
 }
 
 // Connect starts the web server. Satisfies the dex.Connector interface.
-func (s *WebServer) Connect(ctx context.Context) (error, *sync.WaitGroup) {
+func (s *WebServer) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	// We'll use the context for market syncers.
 	s.ctx = ctx
 	// Start serving.
 	listener, err := net.Listen("tcp", s.addr)
 	if err != nil {
-		return fmt.Errorf("Can't listen on %s. web server quitting: %v", s.addr, err), nil
+		return nil, fmt.Errorf("Can't listen on %s. web server quitting: %v", s.addr, err)
 	}
 
 	// Shutdown the server on context cancellation.
@@ -305,7 +305,7 @@ func (s *WebServer) Connect(ctx context.Context) (error, *sync.WaitGroup) {
 	}()
 
 	log.Infof("Web server listening on http://%s", s.addr)
-	return nil, &wg
+	return &wg, nil
 }
 
 // authorize creates, stores, and returns a new auth token to identify the

--- a/client/webserver/webserver_test.go
+++ b/client/webserver/webserver_test.go
@@ -316,7 +316,7 @@ func TestLoadMarket(t *testing.T) {
 	link := newLink()
 	s, tCore, shutdown := newTServer(t, false)
 	defer shutdown()
-	err, _ := link.cl.Connect(tCtx)
+	_, err := link.cl.Connect(tCtx)
 	if err != nil {
 		t.Fatalf("WSLink Start: %v", err)
 	}

--- a/client/webserver/webserver_test.go
+++ b/client/webserver/webserver_test.go
@@ -261,11 +261,14 @@ func newTServer(t *testing.T, start bool) (*WebServer, *TCore, func()) {
 	}
 
 	if start {
-		waiter := dex.NewStartStopWaiter(s)
-		waiter.Start(ctx)
+		cm := dex.NewConnectionMaster(s)
+		err := cm.Connect(ctx)
+		if err != nil {
+			t.Fatalf("Error starting WebServer: %v", err)
+		}
 		shutdown = func() {
 			killCtx()
-			waiter.WaitForShutdown()
+			cm.Disconnect()
 		}
 	} else {
 		shutdown = killCtx
@@ -313,7 +316,10 @@ func TestLoadMarket(t *testing.T) {
 	link := newLink()
 	s, tCore, shutdown := newTServer(t, false)
 	defer shutdown()
-	link.cl.Start()
+	err, _ := link.cl.Connect(tCtx)
+	if err != nil {
+		t.Fatalf("WSLink Start: %v", err)
+	}
 	defer link.cl.Disconnect()
 	params := &marketLoad{
 		Host:  "abc",

--- a/client/webserver/websocket.go
+++ b/client/webserver/websocket.go
@@ -91,8 +91,13 @@ func (s *WebServer) websocketHandler(conn ws.Connection, ip string) {
 		delete(s.clients, cl.cid)
 		s.mtx.Unlock()
 	}()
-	cl.Start()
-	cl.WaitForShutdown()
+	cm := dex.NewConnectionMaster(cl)
+	err := cm.Connect(s.ctx)
+	if err != nil {
+		log.Errorf("websocketHandler client Connect: %v")
+		return
+	}
+	cm.Wait()
 	log.Tracef("Disconnected websocket client %s", ip)
 }
 

--- a/dex/runner.go
+++ b/dex/runner.go
@@ -75,7 +75,7 @@ func (ssw *StartStopWaiter) Stop() {
 // Connector is any type that implements the Connect method, which will return
 // a connection error, and a WaitGroup that can be waited on at Disconnection.
 type Connector interface {
-	Connect(ctx context.Context) (error, *sync.WaitGroup)
+	Connect(ctx context.Context) (*sync.WaitGroup, error)
 }
 
 // ConnectionMaster manages a Connector.
@@ -97,7 +97,7 @@ func NewConnectionMaster(c Connector) *ConnectionMaster {
 func (c *ConnectionMaster) Connect(ctx context.Context) (err error) {
 	c.init(ctx)
 	c.mtx.Lock()
-	err, c.wg = c.connector.Connect(c.ctx)
+	c.wg, err = c.connector.Connect(c.ctx)
 	c.mtx.Unlock()
 	return err
 }

--- a/dex/runner.go
+++ b/dex/runner.go
@@ -109,3 +109,8 @@ func (c *ConnectionMaster) Disconnect() {
 	defer c.mtx.RUnlock()
 	c.wg.Wait()
 }
+
+// Wait waits for the the WaitGroup returned by Connect.
+func (c *ConnectionMaster) Wait() {
+	c.wg.Wait()
+}

--- a/dex/ws/wslink.go
+++ b/dex/ws/wslink.go
@@ -4,10 +4,13 @@
 package ws
 
 import (
+	"context"
 	"encoding/json"
 	"errors"
+	"fmt"
 	"net/http"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"decred.org/dcrdex/dex/msgjson"
@@ -56,34 +59,32 @@ type WSLink struct {
 	ip string
 	// conn is the gorilla websocket.Conn, or a stub for testing.
 	conn Connection
-	// quitMtx protects the on flag and closing of the quit channel.
-	quitMtx sync.RWMutex
+	// writeMtx is a mutex to sequence WriteMessage operations.
+	writeMtx sync.Mutex
 	// on is used internally to prevent multiple Close calls on the underlying
 	// connections.
-	on bool
-	// After disconnect, the quit channel will be closed.
-	quit chan struct{}
+	on uint32
+	// quit is used to cancel the Context.
+	quit context.CancelFunc
+	// outChan is used to sequence sent messages.
+	outChan chan []byte
 	// The WSLink has at least 3 goroutines, one for read, one for write, and
 	// one server goroutine to monitor for peer disconnection. The WaitGroup is
 	// used to synchronize cleanup on disconnection.
 	wg sync.WaitGroup
-	// Messages to the peer are routed through the outChan. This ensures
-	// messages are sent in the correct order, and satisfies the thread-safety
-	// requirements of the (*websocket.Conn).WriteMessage.
-	outChan chan []byte
 	// A master message handler.
 	handler func(*msgjson.Message) *msgjson.Error
 	// pingPeriod is how often to ping the peer.
 	pingPeriod time.Duration
+	// sendWG is a WaitGroup to sequence sends with a graceful shutdown.
+	sendWG sync.WaitGroup
 }
 
 // NewWSLink is a constructor for a new WSLink.
 func NewWSLink(addr string, conn Connection, pingPeriod time.Duration, handler func(*msgjson.Message) *msgjson.Error) *WSLink {
 	return &WSLink{
-		on:         true,
 		ip:         addr,
 		conn:       conn,
-		quit:       make(chan struct{}),
 		outChan:    make(chan []byte, outBufferSize),
 		pingPeriod: pingPeriod,
 		handler:    handler,
@@ -100,15 +101,19 @@ func (c *WSLink) Send(msg *msgjson.Message) error {
 		return err
 	}
 
-	// If outChan is blocking (outBufferSize pending messages), many messages
-	// are being sent and not quickly enough. If the connection is actually
-	// down, c.quit will be closed. TODO: Add a default case with error return
-	// or just block until outHandler receives?
-	select {
-	case c.outChan <- b:
-	case <-c.quit:
-		return ErrPeerDisconnected
-	}
+	c.sendWG.Add(1)
+	c.outChan <- b
+	go func() {
+		c.writeMtx.Lock()
+		b := <-c.outChan
+		c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+		err := c.conn.WriteMessage(websocket.TextMessage, b)
+		c.writeMtx.Unlock()
+		c.sendWG.Done()
+		if err != nil {
+			c.Disconnect()
+		}
+	}()
 	return nil
 }
 
@@ -124,57 +129,54 @@ func (c *WSLink) SendError(id uint64, rpcErr *msgjson.Error) {
 	}
 }
 
-// Start begins processing input and output messages.
-func (c *WSLink) Start() {
+// Connect begins processing input and output messages.
+func (c *WSLink) Connect(ctx context.Context) (error, *sync.WaitGroup) {
 	// Set the initial read deadline now that the ping ticker is about to be
 	// started. The pong handler will set subsequent read deadlines. 2x ping
 	// period is a very generous initial pong wait; the readWait provided to
 	// NewConnection could be stored and used here (once) instead.
+	if !atomic.CompareAndSwapUint32(&c.on, 0, 1) {
+		return fmt.Errorf("Attempted to Start a running WSLink"), nil
+	}
+	linkCtx, quit := context.WithCancel(ctx)
+	c.quit = quit
 	err := c.conn.SetReadDeadline(time.Now().Add(c.pingPeriod * 2))
 	if err != nil {
-		log.Errorf("Failed to set initial read deadline for %v: %v", c.ip, err)
-		return
+		return fmt.Errorf("Failed to set initial read deadline for %v: %v", c.ip, err), nil
 	}
 
 	log.Tracef("Starting websocket messaging with peer %s", c.ip)
 	// Start processing input and output.
 	c.wg.Add(2)
-	go c.inHandler()
-	go c.outHandler()
+	go c.inHandler(linkCtx)
+	go c.pingHandler(linkCtx)
+	return nil, &c.wg
 }
 
 // Disconnect closes both the underlying websocket connection and the quit
 // channel.
 func (c *WSLink) Disconnect() {
-	c.quitMtx.Lock()
-	defer c.quitMtx.Unlock()
-	if !c.on {
+	if !atomic.CompareAndSwapUint32(&c.on, 1, 0) {
 		log.Debugf("Disconnect attempted on stopped WSLink.")
 		return
 	}
 	log.Tracef("Closing connection with peer %v", c.ip)
-	c.on = false
+	c.quit()
+	c.sendWG.Wait()
 	c.conn.Close()
-	close(c.quit)
-}
-
-// WaitForShutdown blocks until the WSLink goroutines are stopped and the
-// connection is closed.
-func (c *WSLink) WaitForShutdown() {
-	c.wg.Wait()
 }
 
 // inHandler handles all incoming messages for the websocket connection. It must
 // be run as a goroutine.
-func (c *WSLink) inHandler() {
+func (c *WSLink) inHandler(ctx context.Context) {
+	// Ensure the connection is closed.
+	defer c.Disconnect()
+	defer c.wg.Done()
 out:
 	for {
-		// Break out of the loop once the quit channel has been closed.
-		// Use a non-blocking select here so we fall through otherwise.
-		select {
-		case <-c.quit:
+		// Quit when the context is closed.
+		if ctx.Err() != nil {
 			break out
-		default:
 		}
 		// Block until a message is received or an error occurs.
 		_, msgBytes, err := c.conn.ReadMessage()
@@ -205,15 +207,11 @@ out:
 			c.SendError(msg.ID, rpcErr)
 		}
 	}
-	// Ensure the connection is closed.
-	c.Disconnect()
-	c.wg.Done()
 }
 
-// outHandler handles all outgoing messages for the websocket connection.
-// It uses a buffered channel to serialize output messages while allowing the
-// sender to continue running asynchronously.  It must be run as a goroutine.
-func (c *WSLink) outHandler() {
+// pingHandler sends periodic pings to the client.
+func (c *WSLink) pingHandler(ctx context.Context) {
+	defer c.wg.Done()
 	ticker := time.NewTicker(c.pingPeriod)
 	ping := []byte{}
 out:
@@ -221,13 +219,6 @@ out:
 		// Send any messages ready for send until the quit channel is
 		// closed.
 		select {
-		case b := <-c.outChan:
-			c.conn.SetWriteDeadline(time.Now().Add(writeWait))
-			err := c.conn.WriteMessage(websocket.TextMessage, b)
-			if err != nil {
-				c.Disconnect()
-				break out
-			}
 		case <-ticker.C:
 			err := c.conn.WriteControl(websocket.PingMessage, ping, time.Now().Add(writeWait))
 			if err != nil {
@@ -236,30 +227,17 @@ out:
 				log.Debugf("WriteMessage ping error: %v", err)
 				break out
 			}
-		case <-c.quit:
+		case <-ctx.Done():
 			break out
 		}
 	}
 
-	// Drain any wait channels before exiting so nothing is left waiting
-	// around to send.
-cleanup:
-	for {
-		select {
-		case <-c.outChan:
-		default:
-			break cleanup
-		}
-	}
-	c.wg.Done()
 	log.Tracef("Websocket output handler done for peer %s", c.ip)
 }
 
 // Off will return true if the link has disconnected.
 func (c *WSLink) Off() bool {
-	c.quitMtx.RLock()
-	defer c.quitMtx.RUnlock()
-	return !c.on
+	return atomic.LoadUint32(&c.on) == 0
 }
 
 // IP is the peer address passed to the constructor.

--- a/dex/ws/wslink.go
+++ b/dex/ws/wslink.go
@@ -170,8 +170,8 @@ func (c *WSLink) Disconnect() {
 // be run as a goroutine.
 func (c *WSLink) inHandler(ctx context.Context) {
 	// Ensure the connection is closed.
-	defer c.Disconnect()
 	defer c.wg.Done()
+	defer c.Disconnect()
 out:
 	for {
 		// Quit when the context is closed.

--- a/dex/ws/wslink.go
+++ b/dex/ws/wslink.go
@@ -130,19 +130,19 @@ func (c *WSLink) SendError(id uint64, rpcErr *msgjson.Error) {
 }
 
 // Connect begins processing input and output messages.
-func (c *WSLink) Connect(ctx context.Context) (error, *sync.WaitGroup) {
+func (c *WSLink) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	// Set the initial read deadline now that the ping ticker is about to be
 	// started. The pong handler will set subsequent read deadlines. 2x ping
 	// period is a very generous initial pong wait; the readWait provided to
 	// NewConnection could be stored and used here (once) instead.
 	if !atomic.CompareAndSwapUint32(&c.on, 0, 1) {
-		return fmt.Errorf("Attempted to Start a running WSLink"), nil
+		return nil, fmt.Errorf("Attempted to Start a running WSLink")
 	}
 	linkCtx, quit := context.WithCancel(ctx)
 	c.quit = quit
 	err := c.conn.SetReadDeadline(time.Now().Add(c.pingPeriod * 2))
 	if err != nil {
-		return fmt.Errorf("Failed to set initial read deadline for %v: %v", c.ip, err), nil
+		return nil, fmt.Errorf("Failed to set initial read deadline for %v: %v", c.ip, err)
 	}
 
 	log.Tracef("Starting websocket messaging with peer %s", c.ip)
@@ -150,7 +150,7 @@ func (c *WSLink) Connect(ctx context.Context) (error, *sync.WaitGroup) {
 	c.wg.Add(2)
 	go c.inHandler(linkCtx)
 	go c.pingHandler(linkCtx)
-	return nil, &c.wg
+	return &c.wg, nil
 }
 
 // Disconnect closes both the underlying websocket connection and the quit

--- a/server/comms/comms_test.go
+++ b/server/comms/comms_test.go
@@ -25,7 +25,10 @@ import (
 	"github.com/gorilla/websocket"
 )
 
-var testCtx context.Context
+var (
+	tErr    = fmt.Errorf("test error")
+	testCtx context.Context
+)
 
 func newServer() *Server {
 	return &Server{
@@ -61,11 +64,12 @@ func readChannel(t *testing.T, tag string, c chan interface{}) interface{} {
 }
 
 type wsConnStub struct {
-	msg   chan []byte
-	quit  chan struct{}
-	read  int
-	close int
-	recv  chan []byte
+	msg      chan []byte
+	quit     chan struct{}
+	close    int
+	recv     chan []byte
+	writeMtx sync.Mutex
+	writeErr error
 }
 
 func (conn *wsConnStub) addChan() {
@@ -88,6 +92,12 @@ func newWsStub() *wsConnStub {
 	}
 }
 
+func (conn *wsConnStub) setWriteErr(err error) {
+	conn.writeMtx.Lock()
+	conn.writeErr = err
+	conn.writeMtx.Unlock()
+}
+
 // nonEOF can specify a particular error should be returned through ReadMessage.
 var nonEOF = make(chan struct{})
 var pongTrigger = []byte("pong")
@@ -107,16 +117,12 @@ func (conn *wsConnStub) ReadMessage() (int, []byte, error) {
 		close(conn.quit)
 		return 0, nil, fmt.Errorf("test nonEOF error")
 	}
-	conn.read++
 	return 0, b, nil
 }
 
-var (
-	writeErr    string
-	writeErrMtx sync.Mutex
-)
-
 func (conn *wsConnStub) WriteMessage(msgType int, msg []byte) error {
+	conn.writeMtx.Lock()
+	defer conn.writeMtx.Unlock()
 	if msgType == websocket.PingMessage {
 		select {
 		case conn.msg <- pongTrigger:
@@ -128,13 +134,11 @@ func (conn *wsConnStub) WriteMessage(msgType int, msg []byte) error {
 	if conn.recv != nil {
 		conn.recv <- msg
 	}
-	writeErrMtx.Lock()
-	defer writeErrMtx.Unlock()
-	if writeErr == "" {
+	if conn.writeErr == nil {
 		return nil
 	}
-	err := fmt.Errorf(writeErr)
-	writeErr = ""
+	err := conn.writeErr
+	conn.writeErr = nil
 	return err
 }
 
@@ -227,7 +231,10 @@ func TestMain(m *testing.M) {
 	var shutdown func()
 	testCtx, shutdown = context.WithCancel(context.Background())
 	defer shutdown()
-	// UseLogger(slog.NewBackend(os.Stdout).Logger("COMMSTEST"))
+	// logger := slog.NewBackend(os.Stdout).Logger("COMMSTEST")
+	// logger.SetLevel(slog.LevelTrace)
+	// UseLogger(logger)
+	// ws.UseLogger(logger)
 	os.Exit(m.Run())
 }
 
@@ -332,7 +339,10 @@ func TestClientRequests(t *testing.T) {
 		conn = newWsStub()
 		wg.Add(1)
 		needCount := server.clientCount() + 1
-		go server.websocketHandler(&wg, conn, stubAddr)
+		go func() {
+			defer wg.Done()
+			server.websocketHandler(testCtx, conn, stubAddr)
+		}()
 		if !giveItASecond(func() bool {
 			return server.clientCount() == needCount
 		}) {
@@ -341,14 +351,13 @@ func TestClientRequests(t *testing.T) {
 		getClient()
 	}
 	reconnect()
-
 	// Check that the request is parsed as expected.
 	sendToServer("checkrequest", `{"key":"value"}`)
 	readChannel(t, "checkrequest", srvChan)
-
 	// Send invalid params, and make sure the server doesn't pass the message. The
 	// server will not disconnect the client.
 	conn.addChan()
+
 	ensureReplaceFails := func(old, new string) {
 		sendReplace(t, conn, makeReq("checkinvalid", old), old, new)
 		<-conn.recv
@@ -356,6 +365,7 @@ func TestClientRequests(t *testing.T) {
 			t.Fatalf("invalid request passed to handler")
 		}
 	}
+
 	ensureReplaceFails(`{"a":"b"}`, "?")
 	if client.Off() {
 		t.Fatalf("client unexpectedly disconnected after invalid message")
@@ -363,9 +373,7 @@ func TestClientRequests(t *testing.T) {
 
 	// Send the invalid message again, but error out on the server's WriteMessage
 	// attempt. The server should disconnect the client in this case.
-	writeErrMtx.Lock()
-	writeErr = "basic error"
-	writeErrMtx.Unlock()
+	conn.setWriteErr(tErr)
 	waitForShutdown("rpc error", func() {
 		ensureReplaceFails(`{"a":"b"}`, "?")
 	})
@@ -393,9 +401,7 @@ func TestClientRequests(t *testing.T) {
 
 	// Again, but with an WriteMessage error when sending error to client. This
 	// should result in a disconnection.
-	writeErrMtx.Lock()
-	writeErr = "basic error"
-	writeErrMtx.Unlock()
+	conn.setWriteErr(tErr)
 	waitForShutdown("rpc error", func() {
 		sendToServer("nonexistent", "{}")
 		conn.wait(t, "bad path with error")
@@ -473,7 +479,7 @@ func TestClientResponses(t *testing.T) {
 	getClient := func() {
 		encReq, _ := json.Marshal(makeReq("grabclient", `{}`))
 		conn.msg <- encReq
-		client = readChannel(t, "getClient", srvChan).(*wsLink)
+		client = readChannel(t, "grabclient", srvChan).(*wsLink)
 	}
 
 	sendToClient := func(route, payload string, f func(Link, *msgjson.Message), expiration time.Duration, expire func()) uint64 {
@@ -497,7 +503,10 @@ func TestClientResponses(t *testing.T) {
 	reconnect := func() {
 		conn = newWsStub()
 		wg.Add(1)
-		go server.websocketHandler(&wg, conn, stubAddr)
+		go func() {
+			defer wg.Done()
+			server.websocketHandler(testCtx, conn, stubAddr)
+		}()
 		getClient()
 	}
 	reconnect()
@@ -596,7 +605,7 @@ func TestOnline(t *testing.T) {
 
 	keyPath := filepath.Join(tempDir, "rpc.key")
 	certPath := filepath.Join(tempDir, "rpc.cert")
-	pongWait = 50 * time.Millisecond
+	pongWait = 100 * time.Millisecond
 	pingPeriod = (pongWait * 9) / 10
 	server, err := NewServer(&RPCConfig{
 		ListenAddrs: []string{":0"},
@@ -716,8 +725,15 @@ func TestOnline(t *testing.T) {
 	if err != nil {
 		t.Fatalf("banuser send error: %v", err)
 	}
+	// Just for sequencing
 	readChannel(t, "noresponse", banChan)
-	err = readChannel(t, "ok", recv).(error)
+
+	msgB := readChannel(t, "banuser msg", recv).([]byte)
+	if !strings.Contains(string(msgB), "test quarantine") {
+		t.Fatalf("wrong ban message received: %s", string(msgB))
+	}
+
+	err = readChannel(t, "banuser err", recv).(error)
 	if err == nil {
 		t.Fatalf("no read error after ban")
 	}

--- a/server/comms/comms_test.go
+++ b/server/comms/comms_test.go
@@ -34,14 +34,50 @@ func newServer() *Server {
 	}
 }
 
+func giveItASecond(f func() bool) bool {
+	ticker := time.NewTicker(time.Millisecond)
+	timeout := time.NewTimer(time.Second)
+	for {
+		if f() {
+			return true
+		}
+		select {
+		case <-timeout.C:
+			return false
+		default:
+		}
+		<-ticker.C
+	}
+}
+
+func readChannel(t *testing.T, tag string, c chan interface{}) interface{} {
+	select {
+	case i := <-c:
+		return i
+	case <-time.NewTimer(time.Second).C:
+		t.Fatalf("%s: didnt't read channel", tag)
+	}
+	return nil
+}
+
 type wsConnStub struct {
-	msg     chan []byte
-	quit    chan struct{}
-	write   int
-	read    int
-	close   int
-	lastMsg []byte
-	recv    chan []byte
+	msg   chan []byte
+	quit  chan struct{}
+	read  int
+	close int
+	recv  chan []byte
+}
+
+func (conn *wsConnStub) addChan() {
+	conn.recv = make(chan []byte)
+}
+
+func (conn *wsConnStub) wait(t *testing.T, tag string) {
+	select {
+	case <-conn.recv:
+	case <-time.NewTimer(time.Second).C:
+		t.Fatalf("%s - wait timeout", tag)
+	}
 }
 
 func newWsStub() *wsConnStub {
@@ -50,15 +86,6 @@ func newWsStub() *wsConnStub {
 		// recv is nil unless a test wants to receive
 		quit: make(chan struct{}),
 	}
-}
-
-var testMtx sync.RWMutex
-
-// check results under lock.
-func lockedExe(f func()) {
-	testMtx.Lock()
-	defer testMtx.Unlock()
-	f()
 }
 
 // nonEOF can specify a particular error should be returned through ReadMessage.
@@ -77,6 +104,7 @@ func (conn *wsConnStub) ReadMessage() (int, []byte, error) {
 	case <-testCtx.Done():
 		return 0, nil, io.EOF
 	case <-nonEOF:
+		close(conn.quit)
 		return 0, nil, fmt.Errorf("test nonEOF error")
 	}
 	conn.read++
@@ -89,8 +117,6 @@ var (
 )
 
 func (conn *wsConnStub) WriteMessage(msgType int, msg []byte) error {
-	testMtx.Lock()
-	defer testMtx.Unlock()
 	if msgType == websocket.PingMessage {
 		select {
 		case conn.msg <- pongTrigger:
@@ -98,12 +124,10 @@ func (conn *wsConnStub) WriteMessage(msgType int, msg []byte) error {
 		}
 		return nil
 	}
-	conn.lastMsg = msg
 	// Send the message if their is a receiver for the current test.
 	if conn.recv != nil {
 		conn.recv <- msg
 	}
-	conn.write++
 	writeErrMtx.Lock()
 	defer writeErrMtx.Unlock()
 	if writeErr == "" {
@@ -164,7 +188,6 @@ func sendToConn(t *testing.T, conn *wsConnStub, method, msg string) {
 		t.Fatalf("error encoding %s request: %v", method, err)
 	}
 	conn.msg <- encMsg
-	time.Sleep(time.Millisecond * 10)
 }
 
 func sendReplace(t *testing.T, conn *wsConnStub, thing interface{}, old, new string) {
@@ -175,7 +198,6 @@ func sendReplace(t *testing.T, conn *wsConnStub, thing interface{}, old, new str
 	s := string(enc)
 	s = strings.ReplaceAll(s, old, new)
 	conn.msg <- []byte(s)
-	time.Sleep(time.Millisecond)
 }
 
 func newTestDEXClient(addr string, rootCAs *x509.CertPool) (*websocket.Conn, error) {
@@ -233,52 +255,58 @@ func TestRoute_PanicsDoubleRegistry(t *testing.T) {
 // Test the server with a stub for the client connections.
 func TestClientRequests(t *testing.T) {
 	server := newServer()
+	var wg sync.WaitGroup
+	defer func() {
+		server.disconnectClients()
+		wg.Wait()
+	}()
 	var client *wsLink
 	var conn *wsConnStub
 	stubAddr := "testaddr"
 	sendToServer := func(method, msg string) { sendToConn(t, conn, method, msg) }
 
-	clientOn := func() bool {
-		var on bool
-		lockedExe(func() {
-			on = !client.Off()
-		})
-		return on
+	waitForShutdown := func(tag string, f func()) {
+		needCount := server.clientCount() - 1
+		f()
+		if !giveItASecond(func() bool {
+			return server.clientCount() == needCount
+		}) {
+			t.Fatalf("%s: waitForShutdown failed", tag)
+		}
 	}
 
 	// Register all methods before sending any requests.
 	// 'getclient' grabs the server's link.
+	srvChan := make(chan interface{})
 	Route("getclient", func(c Link, _ *msgjson.Message) *msgjson.Error {
-		testMtx.Lock()
-		defer testMtx.Unlock()
-		var ok bool
-		client, ok = c.(*wsLink)
+		client, ok := c.(*wsLink)
 		if !ok {
 			t.Fatalf("failed to assert client type")
 		}
+		srvChan <- client
 		return nil
 	})
+	getClient := func() {
+		encReq, _ := json.Marshal(makeReq("getclient", `{}`))
+		conn.msg <- encReq
+		client = readChannel(t, "getClient", srvChan).(*wsLink)
+	}
+
 	// Check request parses the request to a map of strings.
-	var parsedParams map[string]string
 	Route("checkrequest", func(c Link, msg *msgjson.Message) *msgjson.Error {
-		testMtx.Lock()
-		defer testMtx.Unlock()
-		parsedParams = make(map[string]string)
-		err := json.Unmarshal(msg.Payload, &parsedParams)
-		if err != nil {
-			t.Fatalf("request parse error: %v", err)
+		if string(msg.Payload) != `{"key":"value"}` {
+			t.Fatalf("wrong request: %s", string(msg.Payload))
 		}
 		if client.id != c.ID() {
 			t.Fatalf("client ID mismatch. %d != %d", client.id, c.ID())
 		}
+		srvChan <- nil
 		return nil
 	})
 	// 'checkinvalid' should never be run, since the request has invalid
 	// formatting.
-	passed := false
+	var passed bool
 	Route("checkinvalid", func(_ Link, _ *msgjson.Message) *msgjson.Error {
-		testMtx.Lock()
-		defer testMtx.Unlock()
 		passed = true
 		return nil
 	})
@@ -300,53 +328,36 @@ func TestClientRequests(t *testing.T) {
 
 	// A helper function to reconnect to the server and grab the server's
 	// link.
-	var wg sync.WaitGroup
 	reconnect := func() {
 		conn = newWsStub()
 		wg.Add(1)
+		needCount := server.clientCount() + 1
 		go server.websocketHandler(&wg, conn, stubAddr)
-		time.Sleep(time.Millisecond * 10)
-		sendToServer("getclient", `{}`)
+		if !giveItASecond(func() bool {
+			return server.clientCount() == needCount
+		}) {
+			t.Fatalf("failed to add client")
+		}
+		getClient()
 	}
 	reconnect()
 
-	defer func() {
-		server.disconnectClients()
-		wg.Wait()
-	}()
-
-	sendToServer("getclient", `{}`)
-	lockedExe(func() {
-		if client == nil {
-			t.Fatalf("'getclient' failed")
-		}
-		if server.clientCount() != 1 {
-			t.Fatalf("clientCount != 1")
-		}
-	})
-
 	// Check that the request is parsed as expected.
 	sendToServer("checkrequest", `{"key":"value"}`)
-	lockedExe(func() {
-		v, found := parsedParams["key"]
-		if !found {
-			t.Fatalf("Route key not found")
-		}
-		if v != "value" {
-			t.Fatalf(`expected "value", got %s`, v)
-		}
-	})
+	readChannel(t, "checkrequest", srvChan)
 
 	// Send invalid params, and make sure the server doesn't pass the message. The
 	// server will not disconnect the client.
+	conn.addChan()
 	ensureReplaceFails := func(old, new string) {
 		sendReplace(t, conn, makeReq("checkinvalid", old), old, new)
+		<-conn.recv
 		if passed {
 			t.Fatalf("invalid request passed to handler")
 		}
 	}
 	ensureReplaceFails(`{"a":"b"}`, "?")
-	if !clientOn() {
+	if client.Off() {
 		t.Fatalf("client unexpectedly disconnected after invalid message")
 	}
 
@@ -355,36 +366,28 @@ func TestClientRequests(t *testing.T) {
 	writeErrMtx.Lock()
 	writeErr = "basic error"
 	writeErrMtx.Unlock()
-	ensureReplaceFails(`{"a":"b"}`, "?")
-	if clientOn() {
-		t.Fatalf("client connected after WriteMessage error")
-	}
-	if server.clientCount() != 0 {
-		t.Fatalf("clientCount != 0")
-	}
+	waitForShutdown("rpc error", func() {
+		ensureReplaceFails(`{"a":"b"}`, "?")
+	})
 
 	// Shut the client down. Check the on flag.
 	reconnect()
-	lockedExe(func() { client.Disconnect() })
-	time.Sleep(time.Millisecond)
-	if clientOn() {
-		t.Fatalf("shutdown client has on flag set")
-	}
-	// Shut down again.
-	lockedExe(func() { client.Disconnect() })
+	waitForShutdown("flag set", func() {
+		client.Disconnect()
+	})
 
 	// Reconnect and try shutting down with non-EOF error.
 	reconnect()
-	nonEOF <- struct{}{}
-	time.Sleep(time.Millisecond)
-	if clientOn() {
-		t.Fatalf("failed to shutdown on non-EOF error")
-	}
+	waitForShutdown("non-EOF", func() {
+		nonEOF <- struct{}{}
+	})
 
 	// Try a non-existent handler. This should not result in a disconnect.
 	reconnect()
+	conn.addChan()
 	sendToServer("nonexistent", "{}")
-	if !clientOn() {
+	conn.wait(t, "bad path without error")
+	if client.Off() {
 		t.Fatalf("client unexpectedly disconnected after invalid method")
 	}
 
@@ -393,52 +396,51 @@ func TestClientRequests(t *testing.T) {
 	writeErrMtx.Lock()
 	writeErr = "basic error"
 	writeErrMtx.Unlock()
-	sendToServer("nonexistent", "{}")
-	if clientOn() {
-		t.Fatalf("client still connected after WriteMessage error for invalid method")
-	}
+	waitForShutdown("rpc error", func() {
+		sendToServer("nonexistent", "{}")
+		conn.wait(t, "bad path with error")
+	})
 
 	// An RPC error. No disconnect.
 	reconnect()
+	conn.addChan()
 	sendToServer("error", "{}")
-	if !clientOn() {
+	conn.wait(t, "rpc error")
+	if client.Off() {
 		t.Fatalf("client unexpectedly disconnected after rpc error")
 	}
 
 	// Return a user quarantine error.
-	sendToServer("ban", "{}")
-	if clientOn() {
-		t.Fatalf("client still connected after quarantine error")
-	}
+	waitForShutdown("ban", func() {
+		sendToServer("ban", "{}")
+		conn.wait(t, "ban")
+	})
 	if !server.isQuarantined(stubAddr) {
 		t.Fatalf("server has not marked client as quarantined")
 	}
 	// A call to Send should return ErrPeerDisconnected
-	lockedExe(func() {
-		if !errors.Is(client.Send(nil), ws.ErrPeerDisconnected) {
-			t.Fatalf("incorrect error for disconnected client")
-		}
-	})
+	if !errors.Is(client.Send(nil), ws.ErrPeerDisconnected) {
+		t.Fatalf("incorrect error for disconnected client")
+	}
 
 	checkParseError := func() {
-		lockedExe(func() {
-			msg, err := msgjson.DecodeMessage(conn.lastMsg)
-			if err != nil {
-				t.Fatalf("error decoding last message (%s): %v", string(conn.lastMsg), err)
-			}
-			resp, err := msg.Response()
-			if err != nil {
-				t.Fatalf("error decoding response payload: %v", err)
-			}
-			conn.lastMsg = nil
-			if resp.Error == nil || resp.Error.Code != msgjson.RPCParseError {
-				t.Fatalf("no error after invalid id")
-			}
-		})
+		b := <-conn.recv
+		msg, err := msgjson.DecodeMessage(b)
+		if err != nil {
+			t.Fatalf("error decoding last message (%s): %v", string(b), err)
+		}
+		resp, err := msg.Response()
+		if err != nil {
+			t.Fatalf("error decoding response payload: %v", err)
+		}
+		if resp.Error == nil || resp.Error.Code != msgjson.RPCParseError {
+			t.Fatalf("no error after invalid id")
+		}
 	}
 
 	// Test an invalid ID.
 	reconnect()
+	conn.addChan()
 	msg := makeReq("getclient", `{}`)
 	msg.ID = 555
 	sendReplace(t, conn, msg, "555", "{}")
@@ -458,32 +460,28 @@ func TestClientResponses(t *testing.T) {
 
 	// Register all methods before sending any requests.
 	// 'getclient' grabs the server's link.
+	srvChan := make(chan interface{})
 	Route("grabclient", func(c Link, _ *msgjson.Message) *msgjson.Error {
-		testMtx.Lock()
-		defer testMtx.Unlock()
-		var ok bool
-		client, ok = c.(*wsLink)
+		client, ok := c.(*wsLink)
 		if !ok {
 			t.Fatalf("failed to assert client type")
 		}
+		srvChan <- client
 		return nil
 	})
 
 	getClient := func() {
 		encReq, _ := json.Marshal(makeReq("grabclient", `{}`))
 		conn.msg <- encReq
-		time.Sleep(time.Millisecond * 10)
+		client = readChannel(t, "getClient", srvChan).(*wsLink)
 	}
 
 	sendToClient := func(route, payload string, f func(Link, *msgjson.Message), expiration time.Duration, expire func()) uint64 {
 		req := makeReq(route, payload)
-		lockedExe(func() {
-			err := client.Request(req, f, expiration, expire)
-			if err != nil {
-				t.Logf("sendToClient error: %v", err)
-			}
-			time.Sleep(time.Millisecond * 10)
-		})
+		err := client.Request(req, f, expiration, expire)
+		if err != nil {
+			t.Logf("sendToClient error: %v", err)
+		}
 		return req.ID
 	}
 
@@ -493,7 +491,6 @@ func TestClientResponses(t *testing.T) {
 			t.Fatalf("error encoding %v (%T) request: %v", id, id, err)
 		}
 		conn.msg <- encResp
-		time.Sleep(time.Millisecond * 10)
 	}
 
 	var wg sync.WaitGroup
@@ -501,7 +498,6 @@ func TestClientResponses(t *testing.T) {
 		conn = newWsStub()
 		wg.Add(1)
 		go server.websocketHandler(&wg, conn, stubAddr)
-		time.Sleep(time.Millisecond * 10)
 		getClient()
 	}
 	reconnect()
@@ -512,10 +508,9 @@ func TestClientResponses(t *testing.T) {
 	}()
 
 	// Test Broadcast
-	conn.recv = make(chan []byte)                    // for WriteMessage in this test
+	conn.addChan()                                   // for WriteMessage in this test
 	server.Broadcast(makeNtfn("someNote", `"blah"`)) // async conn.recv <- msg send
 	msgBytes := <-conn.recv
-	conn.recv = nil // all done with this chan
 	msg, err := msgjson.DecodeMessage(msgBytes)
 	if err != nil {
 		t.Fatalf("error decoding last message: %v", err)
@@ -525,43 +520,33 @@ func TestClientResponses(t *testing.T) {
 	if err != nil {
 		return
 	}
-	conn.lastMsg = nil
 	if note != "blah" {
 		t.Errorf("wrong note: %s", note)
 	}
 
 	// Send a request from the server to the client, setting a flag when the
 	// client responds.
-	responded := make(chan struct{}, 1)
 	id := sendToClient("looptest", `{}`, func(_ Link, _ *msgjson.Message) {
-		responded <- struct{}{}
+		srvChan <- nil
 	}, time.Hour, func() {})
 
 	// Respond to the server
 	respondToServer(id, `{}`)
-	select {
-	case <-responded:
-		// Response received, no problem.
-	default:
-		// No response falls through.
-		t.Fatalf("no response for looptest")
-	}
+	readChannel(t, "looptest", srvChan)
+	<-conn.recv
 
 	checkParseError := func(tag string) {
-		lockedExe(func() {
-			msg, err := msgjson.DecodeMessage(conn.lastMsg)
-			if err != nil {
-				t.Fatalf("error decoding last message (%s): %v", tag, err)
-			}
-			resp, err := msg.Response()
-			if err != nil {
-				t.Fatalf("error decoding response (%s): %v", tag, err)
-			}
-			conn.lastMsg = nil
-			if resp.Error == nil || resp.Error.Code != msgjson.RPCParseError {
-				t.Fatalf("no error after %s", tag)
-			}
-		})
+		msg, err := msgjson.DecodeMessage(<-conn.recv)
+		if err != nil {
+			t.Fatalf("error decoding last message (%s): %v", tag, err)
+		}
+		resp, err := msg.Response()
+		if err != nil {
+			t.Fatalf("error decoding response (%s): %v", tag, err)
+		}
+		if resp.Error == nil || resp.Error.Code != msgjson.RPCParseError {
+			t.Fatalf("no error after %s", tag)
+		}
 	}
 
 	// Test an invalid id type.
@@ -574,57 +559,35 @@ func TestClientResponses(t *testing.T) {
 	checkParseError("invalid payload")
 
 	// check the response handler expiration
-	lockedExe(func() {
-		client.respHandlers = make(map[uint64]*responseHandler)
-	})
+	client.respHandlers = make(map[uint64]*responseHandler)
 	expiredID := sendToClient("expiration", `{}`, func(_ Link, _ *msgjson.Message) {},
 		200*time.Millisecond, func() { t.Log("Expired (good).") })
+	<-conn.recv
 	// The responseHandler map should contain the ntfn ID since expiry has not
 	// yet arrived.
-	lockedExe(func() {
-		client.reqMtx.Lock()
-		defer client.reqMtx.Unlock()
-		_, found := client.respHandlers[expiredID]
-		if !found {
-			t.Fatalf("response handler not found")
-		}
-		if len(client.respHandlers) != 1 {
-			t.Fatalf("expected 1 response handler, found %d", len(client.respHandlers))
-		}
-	})
+	client.reqMtx.Lock()
+	_, found := client.respHandlers[expiredID]
+	if !found {
+		t.Fatalf("response handler not found")
+	}
+	if len(client.respHandlers) != 1 {
+		t.Fatalf("expected 1 response handler, found %d", len(client.respHandlers))
+	}
+	client.reqMtx.Unlock()
 
 	time.Sleep(250 * time.Millisecond) // >> 200ms - 10ms
-	lockedExe(func() {
-		client.reqMtx.Lock()
-		defer client.reqMtx.Unlock()
-		if len(client.respHandlers) != 0 {
-			t.Fatalf("expired response handler not pruned")
-		}
-		_, found := client.respHandlers[expiredID]
-		if found {
-			t.Fatalf("expired response handler still in map")
-		}
-	})
-
-	// immediate expiration
-	expiredID = sendToClient("expiration", `{}`, func(_ Link, _ *msgjson.Message) {},
-		0, func() { t.Log("Expired (good).") })
-	lockedExe(func() {
-		client.reqMtx.Lock()
-		defer client.reqMtx.Unlock()
-		if len(client.respHandlers) != 0 {
-			t.Fatalf("expired response handler not pruned")
-		}
-		_, found := client.respHandlers[expiredID]
-		if found {
-			t.Fatalf("expired response handler still in map")
-		}
-	})
+	client.reqMtx.Lock()
+	if len(client.respHandlers) != 0 {
+		t.Fatalf("expired response handler not pruned")
+	}
+	_, found = client.respHandlers[expiredID]
+	if found {
+		t.Fatalf("expired response handler still in map")
+	}
+	client.reqMtx.Unlock()
 }
 
 func TestOnline(t *testing.T) {
-	portAddr := ":57623"
-	address := "wss://127.0.0.1:57623/ws"
 	tempDir, err := ioutil.TempDir("", "example")
 	if err != nil {
 		t.Fatalf("TempDir error: %v", err)
@@ -636,19 +599,16 @@ func TestOnline(t *testing.T) {
 	pongWait = 50 * time.Millisecond
 	pingPeriod = (pongWait * 9) / 10
 	server, err := NewServer(&RPCConfig{
-		ListenAddrs: []string{portAddr},
+		ListenAddrs: []string{":0"},
 		RPCKey:      keyPath,
 		RPCCert:     certPath,
 	})
 	if err != nil {
 		t.Fatalf("server constructor error: %v", err)
 	}
+	address := "wss://" + server.listeners[0].Addr().String() + "/ws"
 
 	// Register routes before starting server.
-	// No response simulates a route that returns no response.
-	Route("noresponse", func(_ Link, _ *msgjson.Message) *msgjson.Error {
-		return nil
-	})
 	// The 'ok' route returns an affirmative response.
 	type okresult struct {
 		OK bool `json:"ok"`
@@ -665,6 +625,7 @@ func TestOnline(t *testing.T) {
 		return nil
 	})
 	// The 'banuser' route quarantines the user.
+	banChan := make(chan interface{})
 	Route("banuser", func(c Link, req *msgjson.Message) *msgjson.Error {
 		rpcErr := msgjson.NewError(msgjson.RPCQuarantineClient, "test quarantine")
 		msg, _ := msgjson.NewResponse(req.ID, nil, rpcErr)
@@ -673,6 +634,7 @@ func TestOnline(t *testing.T) {
 			t.Fatalf("banuser route send error: %v", err)
 		}
 		c.Banish()
+		banChan <- nil
 		return nil
 	})
 
@@ -705,24 +667,15 @@ func TestOnline(t *testing.T) {
 		t.Fatalf("remoteClient constructor error: %v", err)
 	}
 
-	var response, r []byte
-	// Check if the response is nil.
-	nilResponse := func() bool {
-		var isNil bool
-		lockedExe(func() { isNil = response == nil })
-		return isNil
-	}
-
 	// A loop to grab responses from the server.
-	var readErr, re error
+	recv := make(chan interface{})
 	go func() {
 		for {
-			_, r, re = remoteClient.ReadMessage()
-			lockedExe(func() {
-				response = r
-				readErr = re
-			})
-			if re != nil {
+			_, r, err := remoteClient.ReadMessage()
+			if err == nil {
+				recv <- r
+			} else {
+				recv <- err
 				break
 			}
 		}
@@ -734,35 +687,23 @@ func TestOnline(t *testing.T) {
 			t.Fatalf("error encoding %s request: %v", route, err)
 		}
 		err = remoteClient.WriteMessage(websocket.TextMessage, b)
-		time.Sleep(time.Millisecond * 10)
 		return err
 	}
 
 	// Sleep for a few  pongs to make sure the client doesn't disconnect.
 	time.Sleep(pongWait * 3)
 
-	err = sendToDEX("noresponse", "{}")
-	if err != nil {
-		t.Fatalf("noresponse send error: %v", err)
-	}
-	if !nilResponse() {
-		t.Fatalf("response set for 'noresponse' request")
-	}
-
+	// Positive path.
 	err = sendToDEX("ok", "{}")
 	if err != nil {
 		t.Fatalf("noresponse send error: %v", err)
 	}
-	if nilResponse() {
-		t.Fatalf("no response set for 'ok' request")
-	}
-	var resp *msgjson.ResponsePayload
-	lockedExe(func() {
-		msg, _ := msgjson.DecodeMessage(response)
-		resp, _ = msg.Response()
-	})
+	b := readChannel(t, "ok", recv).([]byte)
+
+	msg, _ := msgjson.DecodeMessage(b)
+
 	ok := new(okresult)
-	err = json.Unmarshal(resp.Result, ok)
+	err = msg.UnmarshalResult(ok)
 	if err != nil {
 		t.Fatalf("'ok' response unmarshal error: %v", err)
 	}
@@ -775,35 +716,36 @@ func TestOnline(t *testing.T) {
 	if err != nil {
 		t.Fatalf("banuser send error: %v", err)
 	}
-	lockedExe(func() {
-		if readErr == nil {
-			t.Fatalf("no read error after ban")
-		}
-	})
+	readChannel(t, "noresponse", banChan)
+	err = readChannel(t, "ok", recv).(error)
+	if err == nil {
+		t.Fatalf("no read error after ban")
+	}
+
 	// Try connecting, and make sure there is an error.
 	_, err = newTestDEXClient(address, rootCAs)
 	if err == nil {
 		t.Fatalf("no websocket connection error after ban")
 	}
 	// Manually set the ban time.
-	func() {
-		server.banMtx.Lock()
-		defer server.banMtx.Unlock()
-		if len(server.quarantine) != 1 {
-			t.Fatalf("unexpected number of quarantined IPs")
-		}
-		for ip := range server.quarantine {
-			server.quarantine[ip] = time.Now()
-		}
-	}()
+	server.banMtx.Lock()
+	if len(server.quarantine) != 1 {
+		t.Fatalf("unexpected number of quarantined IPs")
+	}
+	for ip := range server.quarantine {
+		server.quarantine[ip] = time.Now()
+	}
+	server.banMtx.Unlock()
 	// Now try again. Should connect.
 	conn, err := newTestDEXClient(address, rootCAs)
 	if err != nil {
 		t.Fatalf("error connecting on expired ban")
 	}
-	time.Sleep(time.Millisecond * 10)
-	clientCount := server.clientCount()
-	if clientCount != 1 {
+	var clientCount uint64
+	if !giveItASecond(func() bool {
+		clientCount = server.clientCount()
+		return clientCount == 1
+	}) {
 		t.Fatalf("server claiming %d clients. Expected 1", clientCount)
 	}
 	conn.Close()

--- a/server/comms/server.go
+++ b/server/comms/server.go
@@ -348,7 +348,8 @@ func (s *Server) disconnectClients() {
 	s.clientMtx.Unlock()
 }
 
-// addClient assigns the client an ID and adds it to the map.
+// addClient assigns the client an ID, adds it to the map, and attempts to
+// connect.
 func (s *Server) addClient(client *wsLink, ctx context.Context) (*dex.ConnectionMaster, error) {
 	s.clientMtx.Lock()
 	defer s.clientMtx.Unlock()

--- a/server/comms/server.go
+++ b/server/comms/server.go
@@ -18,6 +18,7 @@ import (
 	"sync/atomic"
 	"time"
 
+	"decred.org/dcrdex/dex"
 	"decred.org/dcrdex/dex/msgjson"
 	"decred.org/dcrdex/dex/ws"
 	"github.com/decred/dcrd/certgen"
@@ -226,7 +227,10 @@ func (s *Server) Run(ctx context.Context) {
 		// connections. We must wait on each websocketHandler to return in
 		// response to disconnectClients.
 		wg.Add(1)
-		go s.websocketHandler(&wg, wsConn, ip)
+		go func() {
+			defer wg.Done()
+			s.websocketHandler(ctx, wsConn, ip)
+		}()
 	})
 
 	// Start serving.
@@ -290,20 +294,24 @@ func (s *Server) banish(ip string) {
 // websocketHandler handles a new websocket client by creating a new wsClient,
 // starting it, and blocking until the connection closes. This method should be
 // run as a goroutine.
-func (s *Server) websocketHandler(wg *sync.WaitGroup, conn ws.Connection, ip string) {
-	defer wg.Done()
+func (s *Server) websocketHandler(ctx context.Context, conn ws.Connection, ip string) {
 	log.Debugf("New websocket client %s", ip)
 
 	// Create a new websocket client to handle the new websocket connection
 	// and wait for it to shutdown.  Once it has shutdown (and hence
 	// disconnected), remove it.
 	client := newWSLink(ip, conn)
-	s.addClient(client)
-	client.Start()
+	cm, err := s.addClient(client, ctx)
+	if err != nil {
+		log.Errorf("Failed to add client %s", ip)
+		return
+	}
+	defer s.removeClient(client.id)
+
 	// The connection remains until the connection is lost or the link's
 	// disconnect method is called (e.g. via disconnectClients).
-	client.WaitForShutdown()
-	s.removeClient(client.id)
+	cm.Wait()
+
 	// If the ban flag is set, quarantine the client's IP address.
 	if client.ban {
 		s.banish(client.IP())
@@ -341,12 +349,14 @@ func (s *Server) disconnectClients() {
 }
 
 // addClient assigns the client an ID and adds it to the map.
-func (s *Server) addClient(client *wsLink) {
+func (s *Server) addClient(client *wsLink, ctx context.Context) (*dex.ConnectionMaster, error) {
 	s.clientMtx.Lock()
+	defer s.clientMtx.Unlock()
 	client.id = s.counter
 	s.counter++
 	s.clients[client.id] = client
-	s.clientMtx.Unlock()
+	cm := dex.NewConnectionMaster(client)
+	return cm, cm.Connect(ctx)
 }
 
 // Remove the client from the map.

